### PR TITLE
[ios] Make expo-random be an expo module

### PR DESCRIPTION
**Pull request by tsapeta**
# Why

#15847 let us do synchronous functions in Expo modules, so we can finally migrate `expo-random` (at least on iOS)! 🥳 

# How

Some standard work when migrating to Expo module:
- Added `expo-module.config.json`
- Changed the pod name from `EXRandom` -> `ExpoRandom`
- Reimplemented Objective-C module to Swift
- Removed Objective-C code and dummy xcodeproj so it won't be autolinked by RN
- Updated `ExpoRandom.ts` to use `requireNativeModule` function and fall back to the native module from RN (necessary for Android)
- Slightly improved tests in test-suite

# Test Plan

Tested in NCL and test-suite. Also ensured that it works when the remote debugger is enabled.

### CODE REVIEW

File: packages/expo-random/ios/RandomModule.swift [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  var bytes = [UInt8](repeating: 0, count: length)
+  let status = SecRandomCopyBytes(kSecRandomDefault, length, &bytes)
+
+  guard status == errSecSuccess else {
+    throw FailedGeneratingRandomBytesException(status)
+  }
+  return Data(bytes).base64EncodedString()
+}
+
+private class FailedGeneratingRandomBytesException: GenericException<OSStatus> {
```
Comment (EvanBacon) [inline packages/expo-random/ios/RandomModule.swift:27]: Interesting. Maybe add a comment about what `GenericException<OSStatus>` is and its importance in regards to rejections.

File: packages/expo-random/ios/ExpoRandom.podspec [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+  s.dependency 'ExpoModulesCore'
+
+  # Swift/Objective-C compatibility
+  s.pod_target_xcconfig = {
+    'DEFINES_MODULE' => 'YES',
+    'SWIFT_COMPILATION_MODE' => 'wholemodule'
+  }
+
+  if !$ExpoUseSources&.include?(package['name']) && ENV['EXPO_USE_SOURCE'].to_i == 0 && File.exist?("#{s.name}.xcframework") && Gem::Version.new(Pod::VERSION) >= Gem::Version.new('1.10.0')
```
Comment (EvanBacon) [inline packages/expo-random/ios/ExpoRandom.podspec:26]: Is there a list of modules that get packed as frameworks? If so, maybe this module should be added to it.

**Review by EvanBacon — APPROVED** [review]

File: packages/expo-random/ios/ExpoRandom.podspec (hunk 2 above)
Comment (tsapeta) [reply packages/expo-random/ios/ExpoRandom.podspec:26]: Yeah, there is such list here https://github.com/expo/expo/blob/7331a90e889774b99b33b86922dd4ee71817ecab/tools/src/prebuilds/Prebuilder.ts#L21

We can consider adding it there, but it's pretty small lib so I guess there will be no gains of that.

File: packages/expo-random/ios/RandomModule.swift (hunk 1 above)
Comment (tsapeta) [reply packages/expo-random/ios/RandomModule.swift:27]: This is something from the core, it has some doc comments there.
I'm going to add the exceptions concept to the docs as well, but there are still some missing parts that I'd like to implement first.

**Review by bbarthec — APPROVED** [review]

**Review by lukmccall — APPROVED** [review]

**Pull request merged**